architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,469 @@
+* **Strongly typed and support versioning**: recipes support versioning through the use of OCI-compliant registries. They are not strongly types in the sense of having compile-time validation, but they can be validated at the time they are registered as having the correct input parameters and output properties.
+
+### User scenarios
+
+#### Configure a non-Kubernetes Radius environment
+
+As a platform engineer I can initialize a new Radius environment that is connected to a non-Kubernetes compute platform so that Radius can authenticate and deploy to non-Kubernetes platforms.
+
+Changes include:
+
+* New versions of the affected resource types. The old types can still be supported until we decide to remove them.
+* `kind` can be any string that identifies a platform. If set to `kubernetes` the existing logic will stay in place.
+* `namespace` will only be required for "kubernetes" compute kind. It will be deprecated or removed if we convert the Kubernetes deployments to a recipe.
+
+```diff
++resource environment 'Applications.Core/environments@2025-05-01-preview' = {
+  name: 'myenv'
+  properties: {
+-    compute: {
+-      kind: 'kubernetes'
+-      namespace: 'default'
+-      }
+-    }
+    recipeConfig: {
+      env {
+        foo: 'bar'
+      }
+    }
+    recipes: {
+      'Applications.Datastores/redisCaches':{
+        default: {
+          templateKind: 'bicep'
+          plainHttp: true
+          templatePath: 'ghcr.io/radius-project/recipes/azure/rediscaches:latest'
+        }
+      }
+    }
+-   extensions: [
+-      {
+-        kind: 'kubernetesMetadata'
+-        labels: {
+-          'team.contact.name': 'frontend'
+-        }
+-      }
+-    ]
+  }
+}
+```
+
+> NOTE: The resource types shown here are new versions, e.g., `environments@2025-05-01-preview`, not changes to the existing versions of the resource types. This allows us to add recipe support to the core types without breaking existing deployments.
+
+The `compute` platform is removed so that we do not have hard-coded support for specific platforms. The functionality enabled by the `compute.identity` property would be implemented via recipes so that we do not need hard-coded support for specific platforms. We could consider adding a recipe to the `environment` type if there are features enabled by `compute` that we could not achieve using recipes on the other core types.
+
+#### Initialize a workspace
+
+As a platform engineer I can use `rad init` (or the equivalent rad `workspace`, `group` and `environment` commands), plus `rad recipe register`, to set up a non-Kubernetes compute platform.
+
+`rad init` would be unchanged. Platform engineers would have to add recipes for each core type and UDT they plan to use.
```

Comment:
We'd need to edit the default environment set up during `rad init`. Right now, this defaults to a Kubernetes compute environment

Reply:
I am thinking of `rad init` as a default fast start setup, so defaulting to Kubernetes is good. Someone can test their deployment on a local cluster using `rad init`. They can later add new environments with new recipes for additional platforms.